lib/protobuf/dsl/typespecs.ex

```diff
@@ -0,0 +1,127 @@
+defmodule Protobuf.DSL.Typespecs do
+  @moduledoc false
+
+  alias Protobuf.{FieldProps, MessageProps}
+
+  @spec quoted_enum_typespec(MessageProps.t()) :: Macro.t()
+  def quoted_enum_typespec(%MessageProps{field_props: field_props}) do
+    atom_specs =
+      field_props
+      |> Enum.map(fn {_fnum, %FieldProps{name_atom: name}} -> name end)
+      |> union_specs()
+
+    quote do
+      integer() | unquote(atom_specs)
+    end
+  end
+
+  @spec quoted_message_typespec(MessageProps.t()) :: Macro.t()
+  def quoted_message_typespec(%MessageProps{} = message_props) do
+    regular_fields =
+      for {_fnum, prop} <- message_props.field_props,
+          is_nil(prop.oneof),
+          do: {prop.name_atom, field_prop_to_spec(prop)}
+
+    oneof_fields =
+      for {field_name, fnum} <- message_props.oneof do
+        possible_fields =
+          for {_fnum, prop} <- message_props.field_props,
+              prop.oneof == fnum,
+              do: prop
+
+        {field_name, oneof_spec(possible_fields)}
+      end
+
+    extension_fields =
+      if is_list(message_props.extension_range) and message_props.extension_range != [] do
+        [{:__pb_extensions__, quote(do: map())}]
+      else
+        []
+      end
+
+    quote do
+      %__MODULE__{unquote_splicing(regular_fields ++ oneof_fields ++ extension_fields)}
+    end
+  end
+
+  defp oneof_spec(possible_oneof_fields) do
+    possible_oneof_fields
+    |> Enum.map(fn prop ->
+      quote do: {unquote(prop.name_atom), unquote(field_prop_to_spec(prop))}
+    end)
+    |> union_specs()
+  end
+
+  defp field_prop_to_spec(%FieldProps{map?: true, type: map_mod} = prop) do
+    if not Code.ensure_loaded?(map_mod) do
+      raise "module #{inspect(map_mod)} was not loaded, but was expected to be since it's used as a map entry"
+    end
+
+    map_props = map_mod.__message_props__()
+
+    key_spec = scalar_type_to_spec(map_props.field_props[map_props.field_tags.key].type)
+    value_prop = map_props.field_props[map_props.field_tags.value]
+
+    value_spec = type_to_spec(value_prop.type, value_prop)
+
+    value_spec = if prop.embedded?, do: quote(do: unquote(value_spec) | nil), else: value_spec
+    quote do: %{optional(unquote(key_spec)) => unquote(value_spec)}
+  end
+
+  defp field_prop_to_spec(%FieldProps{type: type} = prop) do
+    spec = type_to_spec(type, prop)
+
+    cond do
+      prop.repeated? -> quote do: [unquote(spec)]
+      prop.embedded? -> quote do: unquote(spec) | nil
+      true -> spec
+    end
```

Comment:
Manipulating ASTs rather than strings feels way more convenient. And it still allows us to generate formatted code if we eventually change our minds and need to. Huge win.